src/resolver.rs

```diff
@@ -1681,6 +1681,39 @@ impl<'i> TypeAnnotator<'i> {
     /// annotate an expression statement
     fn visit_statement_expression(&mut self, ctx: &VisitorContext, statement: &AstNode) {
         match statement.get_stmt() {
+            AstStatement::This => {
+                let name = match ctx.pou.and_then(|name| self.index.find_pou(name)) {
+                    Some(PouIndexEntry::FunctionBlock { name, .. }) => name,
+                    Some(PouIndexEntry::Method { parent_name: name, .. })
+                        if self.index.find_pou(name).is_some_and(|it| it.is_function_block()) =>
```

Comment:
self-note: Unsure about this if check, but theoretically this should be OK because the validator should find `THIS` usage not in function blocks